CSI Overlay Documentation (#971)
kubevirt/user-guide: docs/storage/csi_overlay.md

```

Ensure cert manager is fully operational
```bash
kubectl wait --for=condition=Available -n cert-manager --timeout=120s --all deployments
```

Next, create the hostpath provisioner namespace:
```bash
kubectl create -f https://raw.githubusercontent.com/kubevirt/hostpath-provisioner-operator/main/deploy/namespace.yaml
```

Now deploy the webhook
```bash
kubectl create -f https://raw.githubusercontent.com/kubevirt/hostpath-provisioner-operator/main/deploy/webhook.yaml -n hostpath-provisioner
```

Then create the operator
```bash
kubectl create -f https://raw.githubusercontent.com/kubevirt/hostpath-provisioner-operator/main/deploy/operator.yaml -n hostpath-provisioner
```

## Modifiy the HPP Custom Resource (CR)
Now that the operator is installed, we need to modify the hostpath provisioner CR to indicate we want to use the provisioner as a storage overlay solution.

We can do so by configuring the `storagePool` to use a `pvcTemplate` which defines a RWX PVC.

```yaml
apiVersion: hostpathprovisioner.kubevirt.io/v1beta1
kind: HostPathProvisioner
metadata:
  name: hostpath-provisioner
spec:
  featureGates: 
    - "Snapshotting"

>   imagePullPolicy: Always

Reviewer: I would use `IfNotPresent` here. Always works well when developing and you are iterating, but you don't always want to pull in a real environment. I believe always will not do anything if the digest is the same, but `IfNotPresent` seems more appropriate.

  storagePools:
    - name: "nfs-backend"
      path: "/nfs-vol"
      snapshotPath: "/nfs-vol/snapshots"
      snapshotProvider: "reflink"
      pvcTemplate:
        accessModes:
          - ReadWriteMany # this must be defined as RWX
        storageClassName: "nfs-csi"
        resources:
          requests:
            storage: 10Gi
  workload:
    nodeSelector:
      kubernetes.io/os: linux
```

> When the HPP Operator sees the storage pool configured this way, it will mount the RWX volume to each of the nodes in the cluster and will then create a new `StorageClass` that will provision new volumes from the storage pool. 

Reviewer: ```suggestion
When the HPP Operator sees the storage pool configured this way, it will mount the RWX volume to each node in the cluster and will then create a new `StorageClass` that will provision new volumes from that storage pool. 
```

> *Note* You can define the name of the new storage class by setting the optional `overlayClassName` field (the default name will be `hpp-overlay`). 

Now when any new PVCs are configured to use this new storage class, their volumes will be created as new directories within the network share. (for this example at `/nfs-vol`)

> Once the CR is modified, we can deploy it to the `hostpath-provisioner` namespace

Reviewer: Maybe instead of asking the user to modify the cr file, we can do something like this:
```bash
  cat <<- END | kubectl apply -n hostpath-provisioner -f -
apiVersion: hostpathprovisioner.kubevirt.io/v1beta1
kind: HostPathProvisioner
metadata:
  name: hostpath-provisioner
spec:
  featureGates: 
    - "Snapshotting"
  imagePullPolicy: IfNotPresent
  storagePools:
    - name: "nfs-backend"
      path: "/nfs-vol"
      snapshotPath: "/nfs-vol/snapshots"
      snapshotProvider: "reflink"
      pvcTemplate:
        accessModes:
          - ReadWriteMany # this must be defined as RWX
        storageClassName: "nfs-csi"
        resources:
          requests:
            storage: 10Gi
  workload:
    nodeSelector:
      kubernetes.io/os: linux
END
```

# CSI Overlay for VM State PVCs

> The VMState PVC was originally introduced to provide persistent storage for a VMs that require the use of TPM or EFI devices: 

Reviewer 2: ```suggestion
The VM State PVC was originally introduced to provide persistent storage for VMs that require the use of TPM or EFI devices: 
```
Is it 'VMState PVC' or 'VM State PVC'? This page uses both but use the latter more often.

> * `spec.domain.devices.tpm.persistent: true`
> * `spec.domain.devices.efi.persistent: true`

Reviewer 2: ```suggestion

* `spec.domain.devices.tpm.persistent: true`
* `spec.domain.devices.efi.persistent: true`
```
Bullet points need a new line between them and the previous para to render properly.

The solution to this is to deploy a modified version of the Hostpath Provisioner (HPP) that will be "overlayed" atop a RWX Filesystem capable storage provider (such as NFS). 

We can achieve this overlay by leveraging HPP's [PVCTemplate storage pool](https://github.com/kubevirt/hostpath-provisioner-operator?tab=readme-ov-file#custom-resource-with-pvctemplate-storage-pool). This allows us to define a storage pool that uses a single RWX shared PVC. The underlying storage provider will create the RWX PVC at a network share location and the HPP operator will mount this volume on each node in the cluster. 
Once the storage pool is set up, any new volumes using this setup will be provisioned by HPP and the resulting PVC will be created as a directory on top of the network share allowing it to be accessed on any node.

This allows VM State PVCs to bypass the minimum PVC size requirements imposed by some storage provisioners and eliminates the need to recreate these PVCs during a VM migration.

Below are the steps to deploy and configure this CSI Overlay.

## Deploy HPP Operator

First install the cert manager
```bash

> kubectl create -f https://github.com/cert-manager/cert-manager/releases/download/v1.7.1/cert-manager.yaml

Reviewer 2: v1.7.1 seems really old. Also should we hard-code the version or just stick to 'latest'?
Also, it doesn't seem to explain why we need cert manager.

Author: yea I agree this is a bit old, however I am just trying to be consistent with what we currently have in our [hpp deployment docs ](https://github.com/kubevirt/hostpath-provisioner-operator?tab=readme-ov-file#how-to-deploy)

Reviewer: Feel free to update the docs in hpp. Or keep it like this, either way works for me.

Author: ack, put up a PR in hpp to update the readme and updated this to use the latest tagged release

```

Ensure cert manager is fully operational
```bash
kubectl wait --for=condition=Available -n cert-manager --timeout=120s --all deployments
```

Get the latest release version

> > *Note* this feature was introduced in v0.25.0

Reviewer 2: ```suggestion
Get the latest release version

!!! note
    This feature was introduced in HPP v0.25.0

```
The admonition needs a line break after it or else it includes the code block. Since we're changing that, we may as well change the admonition to the improved syntax.

```bash
HPP_RELEASE=$(curl -s "https://api.github.com/repos/kubevirt/hostpath-provisioner-operator/releases/latest" | grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/')
```

Next, create the hostpath provisioner namespace:
```bash
kubectl create -f https://github.com/kubevirt/hostpath-provisioner-operator/releases/download/$HPP_RELEASE/namespace.yaml
```

Now deploy the webhook
```bash

> kubectl create -f https://raw.githubusercontent.com/kubevirt/hostpath-provisioner-operator/main/deploy/webhook.yaml -n hostpath-provisioner

Reviewer 2: Any reason why this uses `main` branch instead of the $HPP_RELEASE variable like the others?

```

Then create the operator
```bash
kubectl create -f https://github.com/kubevirt/hostpath-provisioner-operator/releases/download/$HPP_RELEASE/operator.yaml -n hostpath-provisioner
```

Wait for operator to be ready
```bash
kubectl rollout status -n hostpath-provisioner deployment/hostpath-provisioner-operator --timeout=120s
```

## Deploy the HPP Custom Resource (CR)
Now that the operator is installed, we need to deploy the hostpath provisioner CR to indicate we want to use the provisioner as a storage overlay solution.

We can do so by configuring the `storagePool` to use a `pvcTemplate` which defines a RWX PVC.

```bash
cat <<- END | kubectl apply -n hostpath-provisioner -f -
apiVersion: hostpathprovisioner.kubevirt.io/v1beta1
kind: HostPathProvisioner
metadata:
  name: hostpath-provisioner
spec:
  featureGates: 
    - "Snapshotting"
  imagePullPolicy: IfNotPresent
  storagePools:
    - name: "nfs-backend"
      path: "/nfs-vol"

>       snapshotProvider: "reflink"

Author: also, maybe this is a bit preemptive since I only posted the PR for the api change last night, but I changed this example CR to not include a `snapshotPath` any more since it was not being used and is being removed here 

- https://github.com/kubevirt/hostpath-provisioner-operator/pull/676

      pvcTemplate:
        accessModes:

>           - ReadWriteMany # this must be defined as RWX

Reviewer 2: Maybe instead of the comment we could add a list of prerequisites before the procedure

        storageClassName: "nfs-csi"
        resources:
          requests:
            storage: 10Gi
  workload:
    nodeSelector:
      kubernetes.io/os: linux
END
```

Wait for hotpath provisioner to be available
```bash
kubectl wait hostpathprovisioners.hostpathprovisioner.kubevirt.io/hostpath-provisioner --for=condition=Available --timeout=480s
```

When the HPP Operator sees the storage pool configured this way, it will mount the RWX volume to each node in the cluster and will then create a new `StorageClass` that will provision new volumes from that storage pool. 

> > *Note* You can define the name of the new storage class by setting the optional `overlayClassName` field (the default name will be `hpp-overlay`). 

Reviewer 2: We should probably include an example yaml that shows this field `overlayClassName` in context. If so, it might be worthwhile breaking it out into it's own section. It needn't be big, but it seems sufficiently significant.

Now when any new PVCs are configured to use this new storage class, their volumes will be created as new directories within the network share. (for this example at `/nfs-vol`)

> ## Modify Kubevirt CR

Reviewer 2: ```suggestion
## Modify KubeVirt CR
```

> Once our modified HPP CR is deployed and our new overlay storage class is created, we need to set the `vmStateStorageClass` config variable within the Kubevirt CR to use this new overlay storage class. Now any new VMs that require a VM State PVC will employ this new overlay CSI technique.

Reviewer 2: ```suggestion
Once our modified HPP CR is deployed and our new overlay storage class is created, we need to set the `vmStateStorageClass` config variable within the KubeVirt CR to use this new overlay storage class. Now any new VMs that require a VM State PVC will employ this new overlay CSI technique.
```

```bash
kubectl edit kubevirt -n kubevirt
```

```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:
    developerConfiguration:
      featureGates:
        - VMPersistentState
    vmStateStorageClass: hpp-overlay  # NOTE! this name needs to match with newly created overlay storage class .
```

## Snapshot / Restore support

> In order to take snapshots of VMs that use this CSI Overlay for their VMState PVCs, you must deploy a `volumesnapshotclass` to your cluster.

Reviewer 2: ```suggestion
In order to take snapshots of VMs that use this CSI Overlay for their VM State PVCs, you must deploy a `volumesnapshotclass` to your cluster.
```

> *Note* this snapshotclass is set up to point to storagepool named "nfs-backend" which is defined in the example CR above.
```bash
kubectl create -f https://github.com/kubevirt/hostpath-provisioner-operator/releases/download/$HPP_RELEASE/volumesnapshotclass.yaml

> ```

Reviewer 2: ```suggestion
!!! note 
    This snapshotclass is set up to point to storagepool named "nfs-backend" which is defined in the example CR above.
    ```bash
    kubectl create -f https://github.com/kubevirt/hostpath-provisioner-operator/releases/download/$HPP_RELEASE/volumesnapshotclass.yaml
    ```
```

# CSI Overlay for VM State PVCs

The VMState PVC was originally introduced to provide persistent storage for a VMs that require the use of TPM or EFI devices: 
* `spec.domain.devices.tpm.persistent: true`
* `spec.domain.devices.efi.persistent: true`

> VM State PVCs currently only need to store a small amount of data, on the order of ~10MB. However, many storage providers require a minimum size requirement for PVCs varying from 1GB up to 50GB. When this requirement is applied across all VMs in the cluster, you end up with a lot of wasted storage in these VM State PVCs which can cause you to reach your storage and platform limits must faster.

Reviewer: ```suggestion
VM State PVCs currently only need to store a small amount of data, on the order of ~10MB. However, many storage providers require a minimum size requirement for PVCs varying from 1GB up to 50GB. When this requirement is applied across all VMs in the cluster, you end up with a lot of wasted storage in these VM State PVCs which can cause you to reach your storage and platform limits much faster.
```

The solution to this is to deploy a modified version of the Hostpath Provisioner (HPP) that will be "overlayed" atop a RWX Filesystem capable storage provider (such as NFS). 

> > *Note* This approach currently only supports RWX FS storage but will be extended to support RWO and Block volumes in a future release.

Reviewer: Should we really promise this? We investigated it, but I don't think we have a firm commitment to actually implement it. So maybe remove this note.

Now when any new PVCs are configured to use this new storage class, their volumes will be created as new directories within the network share. (for this example at `/nfs-vol`)

## Modify Kubevirt CR

Once our modified HPP CR is deployed and our new overlay storage class is created, we need to set the `vmStateStorageClass` config variable within the Kubevirt CR to use this new overlay storage class. Now any new VMs that require a VM State PVC will employ this new overlay CSI technique.

```bash
kubectl edit kubevirt -n kubevirt
```

```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  configuration:
    developerConfiguration:
      featureGates:
        - VMPersistentState
    vmStateStorageClass: hpp-overlay  # NOTE! this name needs to match with newly created overlay storage class .
```

## Snapshot / Restore support

In order to take snapshots of VMs that use this CSI Overlay for their VMState PVCs, you must deploy a `volumesnapshotclass` to your cluster.

> *Note* this snapshotclass is set up to point to storagepool named "nfs-backend" which is defined in the example CR above.
```bash

> kubectl create -f https://raw.githubusercontent.com/kubevirt/hostpath-provisioner-operator/main/deploy/volumesnapshotclass.yaml

Reviewer: Maybe point this to the volumesnapshotclass of the `$HPP_RELEASE` like the other examples? I double checked and the file is there.
